tikvrpc/interceptor.go

```diff
@@ -0,0 +1,187 @@
+//     return func(target string, req *Request) (*Response, error) {
+//         log.Println("before")
+//         resp, err := next(target, req)
+//         log.Println("after")
+//         return resp, err
+//     }
+// }
+// txn.SetInterceptor(LogInterceptor)
+// ```
+//
+// Or you want to inject some dependent modules:
+// ```
+// func GetLogInterceptor(lg *log.Logger) Interceptor {
+//     return func(next InterceptorFunc) InterceptorFunc {
+//         return func(target string, req *Request) (*Response, error) {
+//             lg.Println("before")
+//             resp, err := next(target, req)
+//             lg.Println("after")
+//             return resp, err
+//         }
+//     }
+// }
+// txn.SetInterceptor(GetLogInterceptor())
+// ```
+type Interceptor func(next InterceptorFunc) InterceptorFunc
+
+// InterceptorFunc is a callable function used to initiate a request to TiKV.
+// It is mainly used as the parameter and return value of Interceptor.
+type InterceptorFunc func(target string, req *Request) (*Response, error)
+
+// InterceptorChain is used to combine multiple Interceptors into one.
+// Multiple interceptors will be executed in the order of link time, but are more
+// similar to the onion model: The earlier the interceptor is executed, the later
+// it will return.
+//
+// We can use InterceptorChain like this:
+// ```
+// func Interceptor1(next InterceptorFunc) InterceptorFunc {
+//     return func(target string, req *Request) (*Response, error) {
+//         fmt.Println("begin-interceptor-1")
+//         defer fmt.Println("end-interceptor-1")
+//         return next(target, req)
+//     }
+// }
+// func Interceptor2(next InterceptorFunc) InterceptorFunc {
+//     return func(target string, req *Request) (*Response, error) {
+//         fmt.Println("begin-interceptor-2")
+//         defer fmt.Println("end-interceptor-2")
+//         return next(target, req)
+//     }
+// }
+// txn.SetInterceptor(NewInterceptorChain().Link(Interceptor1).Link(Interceptor2).Build())
+// ```
+//
+// Then every time an RPC request is initiated, the following text will be printed:
+// ```
+// begin-interceptor-1
+// begin-interceptor-2
+// /* do request & respond here */
+// end-interceptor-2
+// end-interceptor-1
+// ```
+type InterceptorChain struct {
+	chain []Interceptor
+}
+
+// NewInterceptorChain creates an empty InterceptorChain.
+func NewInterceptorChain() *InterceptorChain {
+	return &InterceptorChain{}
+}
+
+// Link is used to link the next Interceptor.
+// Multiple interceptors will be executed in the order of link time.
+func (c *InterceptorChain) Link(it Interceptor) *InterceptorChain {
+	c.chain = append(c.chain, it)
+	return c
+}
+
+// Build merges the previously linked interceptors into one.
+func (c *InterceptorChain) Build() Interceptor {
+	return func(next InterceptorFunc) InterceptorFunc {
+		for n := len(c.chain) - 1; n >= 0; n-- {
+			next = c.chain[n](next)
+		}
+		return next
+	}
+}
```

Comment:
The API can be greatly simplified:

```go
func ChainInterceptors(...Interceptor) Inteceptor {

}
```

Reply:
I additionally implemented the `ChainInterceptors` function based on `InterceptorChain` struct, because the `InterceptorChain` struct may come in handy in some builder modes.